### PR TITLE
Support dynamic prop setting for order subclasses.

### LINKED ISSUE #37697: Data missing from orders/xx/refunds API when HPOS enabled
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

**If WooCommerce HPOS is enabled:** When making a request for an order's refunds through the WooCommerce REST API, eg. orders/123/refunds, the `amount`, `reason`, and `refunded_by` attributes are all set to 0/empty, when there is a value. The values shown under the `meta_data` attribute:

CleanShot 2023-04-13 at 12 27 40@2x

**If WooCommerce HPOS is disabled (using WordPress posts table):** The data is there as expected:

CleanShot 2023-04-13 at 12 29 38@2x
	

### Expected behavior

Refunds API endpoint should return amount/reason/etc. when the data is there when using WooCommerce orders table (HPOS).

### Actual behavior

The data is empty.

### Steps to reproduce

1. Enable WooCommerce HPOS beta.
2. Make an API request to the WooCommerce REST API to get an order's refunds.
3. You'll notice the amount/reason/other data missing.
4. Disable HPOS / Switch back to WordPress Posts table and do #2 again and you'll see the data is there as expected.

### WordPress Environment

-

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by vedanshujain**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Internal props that are still being stored in the metadata, such as `_refund_amount` and `_refund_reason` for refunds was not being initialized properly since it was set after `$object->read` property was set to true. A consequence of this would be that `$refund->get_data()` will not return the value of these props correctly since they will be stored in the changed array instead of the data array.

This PR fixes the behavior by exposing setting the prop dynamically and registering above keys as internal metadata.

Closes #37697 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

With HPOS set as authoritative:
1. Create an order with at least one product. Set the status to completed. Let's say the ID of the order is XX.
2. From the admin, add a refund to the order. Save the order.
3. Fetch the order via rest API like so: `wp-json/wc/v3/orders/XX/refunds`. In the returned refunds list, check the amount and refund reason is set properly, and is not included as the meta key.

```json
[
  {
    "id": 1792,
    "date_created": "2023-07-10T12:21:21",
    "date_created_gmt": "2023-07-10T12:21:21",
    "amount": "2116.36",
    "reason": "did not liked products",
    "refunded_by": 0,
    "refunded_payment": false,
    "meta_data": [
      {
        "id": 4526,
        "key": "test",
        "value": "test_value"
      }
    ],
...
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement

#### Comment 

</details>

### CODE REVIEW

**Comment by rrennick** [comment]
@vedanshujain Separate from this PR but related to #35022, `refunded_by` is `0`. The CPT data store uses the refund `post_author` to populate refunded_by. But, it does show me as the user who refunded the order in the order edit screen:

Screenshot 2023-07-13 at 3 16 26 PM

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableRefundDataStore.php [hunk 1]
```diff
@@ -80,44 +92,30 @@ public function delete( &$refund, $args = array() ) {
...
-		foreach ( $refunds as $refund ) {
-			$this->set_refund_props( $refund );
-		}
-	}
-
 	/**
 	 * Helper method to set refund props.
 	 *
-	 * @param \WC_Order $refund Refund object.
+	 * @param \WC_Order_Refund $refund Refund object.
```
Comment (rrennick) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableRefundDataStore.php:98]: Needs an `@since`.

**Review by rrennick — APPROVED** [review]
@vedanshujain This tested great. Pre-approving with the note on `@since`.

**Comment by vedanshujain** [comment]
Merging for now, will take a look at `refunded_by` in a separate PR as needed.

**Pull request merged**